block-producer/src/tasks/txqueue/mod.rs

```diff
@@ -0,0 +1,315 @@
+//! The transaction queue takes transactions coming in, validates them, and
+//! eventually sends them out in a batch. We say "sending a batch" to represent
+//! handing over a set of transactions to the batch builder.
+//!
+//! Specifically, the requirements are:
+//! - A transaction that fails validation is dropped
+//! - There are 2 conditions for a batch to be sent:
+//!   1. The internal queue size reaches [`TxQueueOptions::batch_size`]
+//!   2. A transaction in the internal queue has been sitting for more than
+//!      [`TxQueueOptions::tx_max_time_in_queue`]
+
+#[cfg(test)]
+mod tests;
+
+use std::cmp::min;
+use std::sync::Arc;
+use std::{fmt::Debug, time::Duration};
+
+use async_trait::async_trait;
+use miden_objects::transaction::ProvenTransaction;
+use tokio::select;
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
+use tokio::sync::Mutex;
+use tokio::time::{sleep, Sleep};
+
+// TYPE ALIASES
+// ================================================================================================
+
+type SharedProvenTx = Arc<ProvenTransaction>;
+type SharedVec<T> = Arc<Mutex<Vec<T>>>;
+type ReadyQueue = SharedVec<SharedProvenTx>;
+
+// PUBLIC INTERFACE
+// ================================================================================================
+
+/// Contains all the methods for the transaction queue to fetch incoming data.
+#[async_trait]
+pub trait TxQueueHandleIn: Send + Sync + 'static {
+    type ReadTxError: Debug;
+
+    async fn read_transaction(&self) -> Result<ProvenTransaction, Self::ReadTxError>;
+}
+
+/// Contains all the methods for the transaction queue to send messages out.
+#[async_trait]
+pub trait TxQueueHandleOut: Send + Sync + 'static {
+    type VerifyTxError: Debug;
+    type TxVerificationFailureReason: Debug + Send;
```

Comment:
`VerifyTxError` is meant to capture all "unexpected errors" that (probably) make the system halt (e.g. the underlying channel closed). `TxVerificationFailureReason` represents all "expected issues", i.e. actual verification errors; e.g. the transaction is already in the queue.

I documented this in the error's docstring.

`sled.rs`'s [great blog post](https://sled.rs/errors.html) goes into great detail about the benefits of this pattern.

Reply:
Interesting post. To me the question is why propagate the unrecoverable error up the stack? If the goal of that error is to abort why not just `panic!()`?

(It seems the types no longer exist, I'm asking more out of curiosity to understand the rationale).

Reply:
Good question. Because `panic!()`ing right away isn't always what you want to do. Maybe you want to allow the caller (and every other task in the system) to log some things before ultimately failing. Or maybe you want to try to recover: e.g. if you're talking to another task which panics, you could try to restart that task, and try again.

TLDR: when you want to do anything more sophisticated than outright crashing